# <span style="color:rebeccapurple">Python Fundamentals - Part 2</span>

## <br><span style="color:teal">2.0 Part 2: Collections and Loops
Python has several built-in objects for holding other objects (including all the things we may use as **data**). These objects are called **containers** and **collections**. While these two words may mean slightly different things in computer programming, the distinctions are small in Python and they are often used interchangeably.
<br><br>**Loops** provide the power of Python. We can move through the objects in a container, filter based on custom conditions, and perform code on each piece of data exactly how we'd like. Unlike more functional programming languages, Python is designed for messy data (including data that is being inputted by users in real time).
### <br><br>Objects in Part 2
- Lists
- Tuples
- Sets
- Ranges
- Dictionaries

### Functions in Part 2
- List functions
- Dictionary functions

### Concepts in Part 2
- mutable vs immutable objects
- `for` loops
- `if` statements
- basic error handling

## <br><br><span style="color:teal">2.1 Lists

**A list is a collection of objects kept in a strict order. <br>A list is surrounded by square brackets. <br>Items in the list are separated by commas.**

<br><br>A list can contain any type of object:

In [3]:
grades = [93, 75, 80, 98, 100, 64, 88]

*Remember that when we assign variables, nothing is returned.*

In [6]:
students = ["Alex", "Billy", "Casey", "Drake", "Ellis", "Frankie"]

*Notice that using square brackets is all that you need to tell Python that you are creating a list. You do not need to use a function.*

In [9]:
type(students)

list

<br>A list can also contain only one item:

In [12]:
final_grade = [90]

<br>A list can be empty:

In [15]:
new_grades = []

<br>A list can contain a mix of object types:

In [18]:
student_info = ["Alex", 93, "A"]

<br>You can even have a list of lists, called a **nested list**:

In [21]:
gradebook = [["Casey", "Taylor", "Lee"], [93, 75, 80]]

### <br><br><span style="color:red">Exercise: List syntax

Create a new variable called `family` and store the names of your family members as a list (or choose your favorite TV family!). Don't forget that each name should be stored as a string.

In [23]:
family = ["Bob", "Linda", "Tina", "Gene", "Louise"]

In [25]:
print(family)

['Bob', 'Linda', 'Tina', 'Gene', 'Louise']


## <br><br><span style="color:teal">2.2 Mutable vs. immutable objects

Let's learn some list functions. While we're learning list functions, we'll also learn a bit about how the list object works.

As a review, there are two types of functions: 
1. those that take a list as an argument and do something *with* the list
2. methods that follow the list and do something *to* the list

In [28]:
grades = [93, 75, 80, 98, 100, 64, 88]

The `len()` function that we learned with strings also works with lists:

In [31]:
len(grades)

7

<br><br>The `sort()` method:

In [34]:
grades.sort()

Hmm, nothing was returned.

In [37]:
print(grades)

[64, 75, 80, 88, 93, 98, 100]


Uh oh. Our list has changed order. **Unlike strings, list methods can change the list object even if you didn't save the changed object as a new variable.** This is because lists are **mutable** objects, whereas strings, integers, floats, and booleans are **immutable** objects.

<br>*Be careful with list methods, as they will change your object!* We'll talk more about how to get around this later in this lecture. We'll also see why it's useful to have lists be mutable.

<br><br>The `append()` method will add a new item to the end of a list:

In [42]:
students = ["Alex", "Billy", "Casey", "Drake", "Ellis", "Frankie"]

In [44]:
students.append("Gavi")

Just like all objects, you can return the variable value by either just running the name of the variable:

In [47]:
students

['Alex', 'Billy', 'Casey', 'Drake', 'Ellis', 'Frankie', 'Gavi']

Or by printing the variable:

In [50]:
print(students)

['Alex', 'Billy', 'Casey', 'Drake', 'Ellis', 'Frankie', 'Gavi']


<br><br>Regular functions (not methods) will not change your list, because they do something *with* your object, not something *to* your object. The `sum()` function will give you the sum of a list of numbers:

In [53]:
grades = [93, 75, 80, 98, 100, 64, 88]

In [55]:
sum(grades)

598

In [57]:
grades

[93, 75, 80, 98, 100, 64, 88]

### <br><br><span style="color:red">Exercise: List functions

Run the cell below to store the list `colors`. Write code to append a new color of your choice to the list `colors`. Then write code to return the length of the list.

In [61]:
colors = ["navy blue", "gold", "silver", "scarlet"]

In [63]:
colors.append("emerald")

In [65]:
len(colors)

5

## <br><br><span style="color:teal">2.3 More list functions

We are going to use some list functions in this notebook that aren't automatically loaded with Python. They are in a module that is included with Python, called `statistics`.

Run this cell to import the `statistics` module:

In [71]:
import statistics

We already learned some list functions. Let's make a list of banana bunches. Each number in the list represents the number of bananas in one bunch.

In [74]:
bananas_in_bunches = [4, 9, 7, 4, 3, 6]

<br>`sum()`:

In [77]:
total_bananas = sum(bananas_in_bunches)
print(total_bananas)

33


<br>`len()`:

In [80]:
how_many_bunches = len(bananas_in_bunches)
print(how_many_bunches)

6


<br>`append()`:

In [83]:
new_bunch = 8
bananas_in_bunches.append(new_bunch)
print(bananas_in_bunches)

[4, 9, 7, 4, 3, 6, 8]


<br>`sort()`:

In [86]:
bananas_in_bunches.sort()
print(bananas_in_bunches)

[3, 4, 4, 6, 7, 8, 9]


<br>We can also pass a **keyword argument** to the `sort()` function to reverse the sort order of the list:

In [89]:
bananas_in_bunches.sort(reverse=True)
print(bananas_in_bunches)

[9, 8, 7, 6, 4, 4, 3]


As a reminder, we need to pass a **keyword argument** to change a parameter that has a default value. If you want to use something other than the default value, you have to pass the function the keyword with the new value. So for `sort()` the default of the `reverse` argument is `False`. When we want to reverse it, we need to change `reverse` to `True` by passing it as an argument.

<br>**More built-in functions**

In [93]:
smallest_bunch = min(bananas_in_bunches)
print(smallest_bunch)

3


In [95]:
largest_bunch = max(bananas_in_bunches)
print(largest_bunch)

9


<br><br>As a reminder, to use a function from an imported module, you type the name of the module followed by `.` followed by the name of the function. Let's use `statistics.mean()`:

In [98]:
mean_bunch_size = statistics.mean(bananas_in_bunches)
print(mean_bunch_size)

5.857142857142857


<br>`statistics.median()`:

In [101]:
median_bunch_size = statistics.median(bananas_in_bunches)
print(median_bunch_size)

6


### <br><br><span style="color:red">Exercise: Finding list functions

<br>Let's practice finding information online. Use Google (or ChatGPT, if that's already your default search) to search for the correct Python function to calculate standard deviation, and then write the code to find the standard deviation of `bananas_in_bunches`:

In [105]:
bananas_in_bunches = [4, 9, 7, 4, 3, 6]

In [107]:
statistics.stdev(bananas_in_bunches)

2.258317958127243

## <br><br><span style="color:teal">2.4 List operators

We can use comparison operators with lists, just like other objects:

In [112]:
students = ["Alex", "Billy", "Casey", "Drake", "Ellis", "Frankie"]
grades = [93, 75, 80, 98, 100, 64, 88]

In [114]:
len(students) == len(grades)

False

In [116]:
len(students) > len(grades)

False

<br><br>We can use some arithmetic operators with lists:

In [119]:
grades = [93, 75, 80, 98, 100, 64, 88]
more_grades = [70, 93]

In [121]:
grades = grades + more_grades
print(grades)

[93, 75, 80, 98, 100, 64, 88, 70, 93]


In [123]:
grades = grades - more_grades
print(grades)

TypeError: unsupported operand type(s) for -: 'list' and 'list'

<br>You can't subtract lists!

In [126]:
print(grades * 4)

[93, 75, 80, 98, 100, 64, 88, 70, 93, 93, 75, 80, 98, 100, 64, 88, 70, 93, 93, 75, 80, 98, 100, 64, 88, 70, 93, 93, 75, 80, 98, 100, 64, 88, 70, 93]


In [128]:
grades

[93, 75, 80, 98, 100, 64, 88, 70, 93]

## <br><br><span style="color:teal">2.5 List indexing

Indexing works just like it does in strings (the first item is indexed as 0).

In [132]:
students = ["Alex", "Billy", "Casey", "Drake", "Ellis", "Frankie"]

In [134]:
students[0]

'Alex'

We can use negative indexing with lists.

In [137]:
students[-2]

'Ellis'

We can take a sublist from a list.

In [140]:
students[1:4]

['Billy', 'Casey', 'Drake']

<br>We can change the value of individual items in a list using indexing.

In [143]:
students

['Alex', 'Billy', 'Casey', 'Drake', 'Ellis', 'Frankie']

To change Billy to Billie, I can assign the new string "Billie" to "Billy"'s indexed position in the list:

In [146]:
students[1] = "Billie"

In [148]:
students

['Alex', 'Billie', 'Casey', 'Drake', 'Ellis', 'Frankie']

### <br><br><span style="color:red">Exercise: List indexing

Run the code cell below to store the list `states`. Write code to change "Illinois" to "Indiana".

In [152]:
states = ["Hawaii", "Kentucky", "Illinois", "Iowa"]

In [154]:
states[2] = "Indiana"

In [156]:
print(states)

['Hawaii', 'Kentucky', 'Indiana', 'Iowa']


<br><br><br>You can also use a `for loop` to make lots of different changes to a list, which we will learn later.

## <br><br><span style="color:teal">2.6 Indexing multiple levels

You can index characters inside a string inside a list. Like this:

In [162]:
name = ["Colby", "Witherup", "Wood"]
print("My first name starts with " + name[0][0] + ".")

My first name starts with C.


<br>Indexing starts on the outermost level and works its way in. So in the next line of code, the first item indexed, `[2]`, is selecting which item in the list. The second item indexed, `[1:3]`, is selecting the substring of characters from that item.

In [165]:
name[2][1:3]

'oo'

<br><br>You would use the same style of indexing to index items inside a **nested list**:

In [168]:
gradebook = [["Casey", "Maya", "Toni", "Mae"], [85, 95, 100, 88]]

Try to guess what will be returned before you run the following cells.

In [171]:
gradebook[0][-1]

'Mae'

In [173]:
gradebook[1][3]

88

In [175]:
gradebook[0][1][2]

'y'

### <br><br><span style="color:red">Exercise: Indexing Multiple Levels

Write your own name as a list of your first, middle, and last names (or however many names you have):

In [179]:
full_name = ["Colby", "Witherup", "Wood"]

Write code to index the last letter of your first name:

In [ ]:
full_name[0][-1]

<br>You ran an experiment 5 times a day for 3 days. These are your results:

In [182]:
results = [[.001, .455, .789, .765, .602], 
           [.67, .899, .482, .847, .585], 
           [.943, .984, .423, .45, .776]]

<br>What is the value of the very last run on the last day?

In [185]:
results[-1][-1]

0.776

<br>What is the minimum value you obtained on the first day of your experiment?

In [188]:
min(results[0])

0.001

What is the mean of the values you obtained on your second day?

In [191]:
statistics.mean(results[1])

0.6966

## <br><br><span style="color:teal">2.7 More about mutable objects

### <br><br><span style="color:red">Exercise - Code along to learn more about lists

If you want more practice typing code, follow along with me. Otherwise, just watch me code and try to spot what I do differently between example 1 and example 2.

Example 1

In [196]:
frodo = ["hobbit"]

In [198]:
sam = frodo

In [200]:
frodo

['hobbit']

In [202]:
sam

['hobbit']

In [204]:
frodo.append("ring")

In [206]:
frodo

['hobbit', 'ring']

In [208]:
sam

['hobbit', 'ring']

Example 2

In [210]:
harry = ["wizard"]

In [212]:
ron = ["wizard"]

In [214]:
harry

['wizard']

In [216]:
ron

['wizard']

In [218]:
harry.append("scar")

In [220]:
harry

['wizard', 'scar']

In [222]:
ron

['wizard']

<br>**What did we do differently between example 1 and example 2?**

Example 3

In [225]:
frodo = ["hobbit"]

In [227]:
sam = frodo.copy()

In [229]:
frodo

['hobbit']

In [231]:
sam

['hobbit']

In [233]:
frodo.append("ring")

In [235]:
frodo

['hobbit', 'ring']

In [237]:
sam

['hobbit']

## <br><br><span style="color:teal">2.8 Joining items in a list into one string

Sometimes you will need to join the items in a list together into a string.

In [240]:
full_name_list = ["Bartholomew", "JoJo", "Simpson"]

The `join()` function is actually a string method. The string you start with is the string that you want to use to connect all the items in the list. Then you pass the list to the function as an argument:

In [243]:
" ".join(full_name_list)

'Bartholomew JoJo Simpson'

In [245]:
"-".join(full_name_list)

'Bartholomew-JoJo-Simpson'

In [247]:
"DOH".join(full_name_list)

'BartholomewDOHJoJoDOHSimpson'

<br>You can use an empty string to connect all the items in the list with nothing in between:

In [250]:
"".join(full_name_list)

'BartholomewJoJoSimpson'

### <br><br><span style="color:red">Exercise: Joining a list into a string

In [253]:
path_list = ["Documents", "workshops", "bootcamp", "tuesday", "tuesdayLecture.ipynb"]

Use the join function to combine these words into one string. The words should be separated by a "/" to create a file path.

In [255]:
"/".join(path_list)

'Documents/workshops/bootcamp/tuesday/tuesdayLecture.ipynb'

*Bonus exercise*: Join the path_list together with a "\\" for a Windows computer. Does it behave as you expected? Can you think of a solution?

In [258]:
"\\".join(path_list)

'Documents\\workshops\\bootcamp\\tuesday\\tuesdayLecture.ipynb'

In [261]:
print("\\".join(path_list))

Documents\workshops\bootcamp\tuesday\tuesdayLecture.ipynb


## <br><br><span style="color:teal">2.9 Splitting a string into a list

Another common task is splitting a string into a list. This is also a string method, so let's load a string to practice with:

In [267]:
csv_line = "sample 1,5,24,864,NA,.4556,,,65"

You must pass the `split()` function an argument - what string do you want to use to separate the items for your list. Here we'll use a comma:

In [270]:
csv_line_list = csv_line.split(",")
print(csv_line_list)

['sample 1', '5', '24', '864', 'NA', '.4556', '', '', '65']


### <br><br><span style="color:red">Exercise: Splitting a string

In [273]:
my_data = "Year: 2005\nYear: 2007\nYear: 2010\nYear: 2011\nYear: 2015"

Run the cell above. We want to turn the string `my_data` into a list of years, without any other text or new lines.
<br><br>First, write code to remove every occurrence of "Year: " from the string. You'll have to remember a string function you learned yesterday:

In [284]:
my_data_clean = my_data.replace("Year: ", "")
print(my_data_clean)

2005
2007
2010
2011
2015


<br>Now, write code to split the string into a list of years:

In [287]:
year_list = my_data_clean.split("\n")
print(year_list)

['2005', '2007', '2010', '2011', '2015']


<br>*Challenge:* Change `my_data` into a list of years again, but this time do it in only one line of code:

In [290]:
year_list = my_data.replace("Year: ", "").split("\n")
print(year_list)

['2005', '2007', '2010', '2011', '2015']


## <br><br><span style="color:teal">2.10 for loops

Before we start coding with "for loops", let's talk about how they work - **DEMO in the slides**

<br><br>We're going to work with a list of student grades.

In [296]:
grades = [73, 64, 89, 93, 59, 100, 79]

<br>When writing the for loop, the lines **inside** the loop are indented. You can use four spaces or one tab to indent. Jupyter Lab, Google Colab, and many other Python IDEs will automatically change a tab into four spaces. This is because all the indentations in a script have to match - when using **scripts** you can only use all tabs or all spaces or else you'll get an error - so the IDE does the work for you of making everything match as spaces. *However, when writing scripts in a text editor, you'll need to do the work yourself and make sure they all match*. If you ever watched the show Silicon Valley, now you can understand this scene: https://www.youtube.com/watch?v=cowtgmZuai0. 

In [299]:
for grade in grades:
    print(grade)

73
64
89
93
59
100
79


*Notice that the items are printed in the same order as the list.*

<br>We can change the name of our temporary variable:

In [303]:
for g in grades:
    print(g)

73
64
89
93
59
100
79


In [305]:
for sandwich in grades:
    print(sandwich)

73
64
89
93
59
100
79


<br>Let's say we want to give each student an extra 5 points:

In [308]:
for g in grades:
    print(g + 5)

78
69
94
98
64
105
84


### <br><br><span style="color:red">Exercise: for loop syntax

Run the following cell to store a list of coding languages.

In [312]:
languages = ["Python", "R", "JavaScript", "Julia", "Swift", "C#", "PHP", "BASIC", "C++", "Java"]

Write a for loop to print each item in the `languages` list.

*Be prepared for the IDE to automatically indent the line under your for loop.*

In [316]:
for i in languages:
    print(i)

Python
R
JavaScript
Julia
Swift
C#
PHP
BASIC
C++
Java


#### <br><br>Some common loop errors

Can you see the problem with each of these loops?

In [319]:
for g in grades:
    print(g)

73
64
89
93
59
100
79


In [321]:
for g in grades:
    print(g)

73
64
89
93
59
100
79


In [323]:
for g in grades:
    print(g)

73
64
89
93
59
100
79


## <br><br><span style="color:teal">2.11 if statements

These are also called **conditional statements** or even **conditionals**. Try to predict what will happen when you run these code cells:

In [328]:
if 8 > 10:
    print("Wow!")

In [330]:
if 10 > 8:
    print("That seems right.")

That seems right.


In [332]:
if 8 > 10:
    print("Wow!")
else:
    print("Hmmm... That seems wrong.")

Hmmm... That seems wrong.


<br><br>**if statements** are often used with for loops to filter data for different conditions.

In [335]:
grades = [73, 64, 89, 93, 59, 100, 79]

We can use a series of if/elif/else statements to perform different actions on grades in different ranges:

In [338]:
for g in grades:
    if g >= 90:
        print(g)
        print("grade is A")
    elif g >= 80:
        print(g)
        print("grade is B")
    elif g >= 70:
        print(g)
        print("grade is C")
    elif g >= 60:
        print(g)
        print("grade is D")
    else:
        print(g)
        print("grade is Fail")

73
grade is C
64
grade is D
89
grade is B
93
grade is A
59
grade is Fail
100
grade is A
79
grade is C


<br>Here we are going to print only grades that are 60 or higher:

In [341]:
for g in grades:
    if g >= 60:
        print(g)
    else:
        pass

73
64
89
93
100
79


<br>We don't have to explicitly pass if the condition isn't met. **You do not need to include an `else` statement.**

In [344]:
for g in grades:
    if g >= 60:
        print(g)

73
64
89
93
100
79


### <br><br><span style="color:red">Exercise: Writing if statements

Store this list of Therapod dinosaurs (meat eaters that walk on two legs).

In [348]:
therapods = ["Albertosaurus", 
             "Allosaurus", 
             "Baryonyx", 
             "Carnotaurus", 
             "Coelophysis", 
             "Compsognathus", 
             "Deinonychus", 
             "Giganotosaurus", 
             "Megalosaurus", 
             "Ornithomimus", 
             "Oviraptor", 
             "Saurophaganax", 
             "Spinosaurus", 
             "Tyrannosaurus", 
             "Tyrannotitan", 
             "Velociraptor",
             "Yangchuanosaurus"]

*Note: When typing a long list, you can use multiple lines. Try to line up the next line with the first item in the list.*

Loop through the list `therapods`. Use string indexing to check **if** the first letter of the dinosaur's name is equal to T. If it is T, print the dinosaur's name. If it is not T, print "Not T".

In [352]:
for dino in therapods:
    if dino[0] == "T":
        print(dino)
    else:
        print("Not T")

Not T
Not T
Not T
Not T
Not T
Not T
Not T
Not T
Not T
Not T
Not T
Not T
Not T
Tyrannosaurus
Tyrannotitan
Not T
Not T


## <br><br><span style="color:teal">2.12 Breaking a loop 
We can also stop the loop if a condition is or isn't met:

In [356]:
print(grades)

[73, 64, 89, 93, 59, 100, 79]


In [358]:
for g in grades:
    if g >= 70:
        print("This student is doing ok.")
    else:
        print("I give up. I quit.")
        break

This student is doing ok.
I give up. I quit.


<br>Here I'm switching the order of the last two lines. Take a minute to try and predict what will happen before running the code.

In [361]:
for g in grades:
    if g >= 70:
        print("This student is doing ok.")
    else:
        break
        print("I give up. I quit.")

This student is doing ok.


## <br><br><span style="color:teal">2.13 The in operator
#### New boolean operator

So far we've learned `<` `>` `<=` `>=` `==` and `!=`.

`in` is another Boolean operator. It works for both lists and strings. Take a minute to predict if each boolean statement will return True or False before you run the cell.

In [366]:
"pie" in "pizza pie"

True

In [368]:
"Oreo" in ["Chips Ahoy", "Oreo", "Oatmeal raisin"]

True

In [370]:
cookies = ["Chips Ahoy", "Oreo", "Oatmeal raisin"]
"Oreo" in cookies

True

In [372]:
"oreo" in ["Chips Ahoy", "Oreo", "Oatmeal raisin"]

False

In [374]:
"Ahoy" in "Chips Ahoy"

True

In [376]:
"Ahoy" in ["Chips Ahoy", "Oreo", "Oatmeal raisin"]

False

<br>We can use the `in` comparison operator with if statements.

In [379]:
cookies = ["Chips Ahoy", "Oreo", "Oatmeal raisin"]
for c in cookies:
    if "Ahoy" in c:
        print(c)

Chips Ahoy


<br><br>Let's say we want to give one extra point to anyone who is right on the cusp of getting a better grade:

In [382]:
grades = [73, 64, 89, 93, 59, 100, 79]
grades_to_round = [59, 69, 79, 89]

We'll loop through the `grades` list. **If** the grade is **in** the `grades_to_round` list, we'll add one point to the grade. **If** the grade is not **in** `grades_to_round`, we'll print the original grade.

In [385]:
for g in grades:
    if g in grades_to_round:
        print(g + 1)
    else:
        print(g)

73
64
90
93
60
100
80


<br><br>There's also `not in`:

In [388]:
for g in grades:
    if g not in grades_to_round:
        print(g)
    else:
        print(g + 1)

73
64
90
93
60
100
80


<br><br>If you want your code to run faster, you should think about when to use `in` vs. `not in`. If you think the majority of the items in your loop will be `in`, you should put that condition in the `if` statement. If you think the majority of the items in your loop will be `not in`, you should put that condition in the `if` statement. This is because once the condition is met for an item, the computer does not need to read any following `elif` or `else` statements, saving a bit of time.

### <br><br><span style="color:red">Exercise: The `in` boolean

Make a list of the toppings you like to eat on your ideal pizza (don't forget to run the cell):

In [393]:
my_toppings = ["garlic", "pepperoni", "cicadas"]

Run the cell below to store a list of toppings that are available at the store.

In [396]:
store_toppings = ["pepperoni", 
                  "figs", 
                  "olives", 
                  "pistachios", 
                  "jalapenos", 
                  "mozzarella", 
                  "marscapone", 
                  "tomato sauce", 
                  "garlic", 
                  "mushrooms", 
                  "extra cheese",
                  "tomatoes"]

Loop through the `my_toppings` list. If the topping is also in the `store_toppings` list, print out a message saying that the topping is available. If the topping in your list is not available at the store, print out a message telling you that the topping isn't available.

In [399]:
for top in my_toppings:
    if top in store_toppings:
        print(top + " is available.")
    else:
        print(top + " is not available.")

garlic is available.
pepperoni is available.
cicadas is not available.


## <br><br><span style="color:teal">2.14 Basic error handling

**Short DEMO in the slides.** You will get more practice with try/except later.

## <br><br><span style="color:teal">2.15 More list practice

### <br><span style="color:red">Exercise: Indexing lists

In [404]:
months = ["January", "February", "March", "April", 
          "May", "June", "July", "August", "September", 
          "October", "November", "December"]

*Note: When typing a long list, you can use multiple lines. Try to line up the next line with the first item in the list.*

#### <br>Run the line of code above. Use list indexing to answer these questions:

Which month has your favorite holiday?

In [409]:
months[-1]

'December'

Which months do you consider to be part of Summer?

In [412]:
months[5:9]

['June', 'July', 'August', 'September']

What month were you born? Try to print out your answer in a complete sentence that includes your indexed month.

In [415]:
months[6]

'July'

### <br><br><span style="color:red">Exercise: Comparisons between lists

In [418]:
characters = ["Aang", "Katara", "Sokka", "Zuko", "Iroh", "Toph"]
water_tribe = ["Katara", "Sokka"]

Based on the lists above, will the following booleans evaluate to True or False? Try to answer before running the code to check:

In [421]:
len(characters) > len(water_tribe) and len(water_tribe) > 3

False

In [423]:
"Aang" in water_tribe or "Sokka" not in water_tribe

False

In [425]:
characters[3] == water_tribe[1] or characters[0][0] == water_tribe[0][-1]

False

### <br><br><span style="color:red">Exercise: List functions

Check out all of the list method functions at this link:
<br>[https://docs.python.org/3/tutorial/datastructures.html](https://docs.python.org/3/tutorial/datastructures.html)
<br>Choose at least one list method from the website that we haven't learned yet. Create a list of anything to practice with, and then try out your chosen list method function to see how it works.

## <br><br><span style="color:teal">2.16 Adding items to an empty list

**<span style="color:crimson">LOGIC** *This is a very common task in Python.*

<br>We'll start with a list of grades. Then, let's create a new list of grades, giving everyone 5 extra points. Run this cell to save our list of grades:

In [428]:
grades = [73, 64, 89, 93, 59, 100, 79]

First, we create a new **empty list**.

In [431]:
new_grades = []

Then we loop through our original list and **append** the changed items to our new list.

In [434]:
for g in grades:
    new_grades.append(g + 5)

In [436]:
print(new_grades)

[78, 69, 94, 98, 64, 105, 84]


<br>If you are working with more complicated calculations, you will want to be more **explicit** by defining a new variable inside the loop instead of doing the calculations inside the `append()` function:

In [439]:
new_grades = []

In [441]:
for g in grades:
    new_g = round((g + 5) / 100, 2)
    new_grades.append(new_g)

In [443]:
print(new_grades)

[0.78, 0.69, 0.94, 0.98, 0.64, 1.05, 0.84]


<br>There's one more thing we should consider. Go up and rerun the code cell with the `for loop` one more time. Then run the cell where we print the value of `new_grades`. Do this a few times and you'll see that there's a problem. Every time we rerun the loop, it's just adding the new grades to that same list.
<br><br>Because we're not recreating the empty list in the same code cell as the `for loop`, it's easy to forget to rerun the `new_grades = []` cell every time we run our loop. Especially when we're testing out our code. To prevent this, we should include the creation of the empty list in the same code cell as the `for loop`, like this:

In [446]:
new_grades = []
for g in grades:
    new_g = round((g + 5) / 100, 2)
    new_grades.append(new_g)

In [448]:
print(new_grades)

[0.78, 0.69, 0.94, 0.98, 0.64, 1.05, 0.84]


### <br><br><span style="color:red">Exercise: Adding items to an empty list

Store this list of monthly bills that you pay:

In [452]:
bills = [115.46, 70.40, 26.90, 170.83, 1250.00, 65.40]

First, create a new empty list called `rounded_bills`. Then, loop through the list `bills`. **For** each item in the list, **round** the item to the closest whole number and **append** it to the list `rounded_bills`.

In [455]:
rounded_bills = []
for b in bills:
    rounded_bills.append(round(b))

In [457]:
print(rounded_bills)

[115, 70, 27, 171, 1250, 65]


## <br><br><span style="color:teal">2.16 More practice with loop logic

<br>We just walked through one common situation you'll use in your own Python code - looping through a list and appending items to a new list. 
<br><br>**<span style="color:crimson">LOGIC** Now we'll work through 4 other common situations you'll encounter. These exercises are about applying logic to solve problems - you already know all the objects, functions, and syntax needed to solve these.

### <br>Example 1 - Doing something with each item in a list

<br>First, let's store a list of the forecasted high temperatures on Northwestern's Qatar campus for this coming week:

In [460]:
c_temps = [33, 35, 39, 37, 38, 36, 36]

<br>The equation to convert from Celsius to Farenheit is F = C x 9/5 + 32. 
<br>We can write a for loop to convert our `c_temps` to Farenheit:

In [463]:
for temp in c_temps:
    f = temp * (9/5) + 32
    print(f)

91.4
95.0
102.2
98.60000000000001
100.4
96.8
96.8


<br>Let's improve this output by rounding the temperatures to one digit past the decimal, and by writing a complete sentence:

In [466]:
for temp in c_temps:
    f = temp * (9/5) + 32
    f2 = round(f, 1)
    print("The high temperature was " + str(f2) + " degrees F (" + str(temp) + " C).")

The high temperature was 91.4 degrees F (33 C).
The high temperature was 95.0 degrees F (35 C).
The high temperature was 102.2 degrees F (39 C).
The high temperature was 98.6 degrees F (37 C).
The high temperature was 100.4 degrees F (38 C).
The high temperature was 96.8 degrees F (36 C).
The high temperature was 96.8 degrees F (36 C).


### <br><span style="color:red">Loops Exercise 1

First, store this list. It contains the world record fastest 5K times for men and women:

In [470]:
fastest5k = [12.817, 13.9]

<br>5 kilometers is equal to 3.10686 miles.

To convert a runner's 5K time to their 1 mile time, use the equation:
<br>mile_time = fiveK_time / 3.10686

<br>Write a for loop to loop through the `fastest5k` list. Inside the for loop, convert the 5K time to a mile time, round the mile time to two digits after the decimal point, and print the answer in a complete sentence.

In [478]:
for time in fastest5k:
    mile = time / 3.10686
    print(str(time) + " 5k time is equivalent to a 1 mile time of " + str(mile) + ".") 

12.817 5k time is equivalent to a 1 mile time of 4.125387046728852.
13.9 5k time is equivalent to a 1 mile time of 4.473970503981512.


### <br><br><br>Example 2 - Using all the items in the list to find one number

This list contains the names of students in your class:

In [484]:
students = ["Eleven", "Dustin", "Mike", "Will", "Lucas", "Max"]

<br>On your classroom wall, you want to spell out all of the students' names with one letter per piece of paper, but you need to know how many pieces of paper you need. Luckily, the `len()` function will tell you how many letters are in each name, but you need to tally up the total.

<br>To do this, you are first going to assign the number zero to a variable:

In [488]:
papers = 0

<br>Next, you can loop through the list, adding the length of each name to your `papers` variable as you go:

In [491]:
for student in students:
    papers = papers + len(student)

Let's see if it worked:

In [494]:
print(papers)

28


<br>This is a common situation that comes up when you're coding, so it's good to practice it now.

<br>Remember the lesson we learned earlier today when we created an empty list in a code cell outside the cell where we wrote our `for loop`? Go up and rerun the cell with our `for loop` and then the cell where we print `papers`.
<br><br>Oops! We forgot to reset `papers` to 0 before rerunning the `for loop`. We should have written the code in one cell:

In [498]:
papers = 0
for student in students:
    papers = papers + len(student)
print(papers)

28


### <br><span style="color:red">Loops Exercise 2

<br>Run the cell below to store a list of birds you observed at the bird feeder outside your window:

In [502]:
bird_obs = ["Mourning Dove", 
            "Red-winged Blackbird", 
            "Red-winged Blackbird", 
            "Red-winged Blackbird", 
            "Mourning Dove", 
            "Mourning Dove", 
            "Northern Cardinal", 
            "Mourning Dove", 
            "Song Sparrow", 
            "Song Sparrow", 
            "Song Sparrow", 
            "American Crow",
            "American Robin", 
            "American Crow",
            "Song Sparrow", 
            "Red-winged Blackbird", 
            "Mourning Dove", 
            "Red-winged Blackbird", 
            "Blue Jay", 
            "Red-winged Blackbird", 
            "Song Sparrow", 
            "Red-winged Blackbird", 
            "Mourning Dove", 
            "Red-winged Blackbird", 
            "European Starling"]

<br>How many Red-winged Blackbirds did you see?

First create a variable called `blackbird_count` that stores the number 0. Then, write a for loop to loop through the `bird_obs` list. *If* the bird is a Red-winged Blackbird, reassign the variable `blackbird_count` to it's current value plus 1.

In [506]:
blackbird_count = 0
for bird in bird_obs:
    if bird == "Red-winged Blackbird":
        blackbird_count = blackbird_count + 1

Now print the variable `blackbird_count`:

In [509]:
print(blackbird_count)

8


### <br><br><br>Example 3 - Using a for loop with if/elif/else statements to filter the items in a list.

In [512]:
students = ["Eleven", "Dustin", "Mike", "Will", "Lucas", "Max"]

In [514]:
for student in students:
    if student == "Will":
        print(student + " should repeat 7th grade because of absences.")
    elif student == "Eleven":
        print("I'm not sure who this student is, they just showed up one day.")
    else:
        print(student + " can move up to 8th grade.")

I'm not sure who this student is, they just showed up one day.
Dustin can move up to 8th grade.
Mike can move up to 8th grade.
Will should repeat 7th grade because of absences.
Lucas can move up to 8th grade.
Max can move up to 8th grade.


<br>We can add more code under each if/elif/else statement. Let's add the students to the correct empty list:

In [517]:
grade_8 = []
grade_7 = []
unsure = []
for student in students:
    if student == "Will":
        print(student + " should repeat 7th grade because of absences.")
        grade_7.append(student)
    elif student == "Eleven":
        print("I'm not sure who this student is, they just showed up one day.")
        unsure.append(student)
    else:
        print(student + " can move up to 8th grade.")
        grade_8.append(student)

I'm not sure who this student is, they just showed up one day.
Dustin can move up to 8th grade.
Mike can move up to 8th grade.
Will should repeat 7th grade because of absences.
Lucas can move up to 8th grade.
Max can move up to 8th grade.


In [519]:
print("Grade 7:")
print(grade_7)
print("Grade 8:")
print(grade_8)
print("Unsure where to place:")
print(unsure)

Grade 7:
['Will']
Grade 8:
['Dustin', 'Mike', 'Lucas', 'Max']
Unsure where to place:
['Eleven']


### <br><span style="color:red">Loops Exercise 3

*This one will take a little thought.*

In [522]:
grades = [90, 75, 92, 89, 95, 94, 100, 92]

The list `grades` shows the homework grades that one student received over the quarter. You told the students that you would drop their one lowest homework grade.

<br>First, create a new empty list called `final_grades`. Then, write a for loop to loop through `grades`. **For** each grade in the list, **if** the grade is **not equal to** the lowest grade in the list, **append** it to `final_grades`:
<br>*Hint: you can use the min() function to find the lowest grade in the list.*

In [524]:
final_grades = []
for g in grades:
    if g != min(grades):
        final_grades.append(g)

Print `final_grades`.

In [527]:
print(final_grades)

[90, 92, 89, 95, 94, 100, 92]


### <br><br><br>Example 4: Using try/except to handle unexpected variations in a list

We often need to use try/except if we aren't sure whether all the items in our list are of the correct data type.

In [531]:
test_scores = [90, 95, "absent", 100, 76]

Let's say we want to give an extra 2 points for every test.

In [534]:
bonus_scores = []
for score in test_scores:
    bonus_scores.append(score + 2)
print(bonus_scores)

TypeError: can only concatenate str (not "int") to str

<br>This gave us an error because you can't add the number 2 to the word "absent", so we can rewrite it:

In [537]:
bonus_scores = []
for score in test_scores:
    try:
        bonus_scores.append(score + 2)
    except TypeError:
        bonus_scores.append(score)
print(bonus_scores)

[92, 97, 'absent', 102, 78]


<br>You need to specify the type of error in your except statement.
<br><br>Sometimes this means that you need to run the code without try/except first to get the name of the error you should expect.
<br><br>If you don't include the error type, it will perform the except code for all exceptions, which might seem great, but it can actually create more trouble for you down the road.

In [540]:
test_scores2 = [90, 95, "absent", 100, "76"]

Here we have the same list, except you accidentally entered the last score as a string instead of an integer. Let's try this list with the same code we just wrote:

In [543]:
bonus_scores2 = []
for score in test_scores2:
    try:
        bonus_scores2.append(score + 2)
    except TypeError:
        bonus_scores2.append(score)
print(bonus_scores2)

[92, 97, 'absent', 102, '76']


<br>It didn't add bonus points to the last score, even though you need that score included. Let's try this to account for the last score:

In [546]:
bonus_scores2 = []
for score in test_scores2:
    try:
        bonus_scores2.append(score + 2)
    except TypeError:
        bonus_scores2.append(int(score) + 2)
print(bonus_scores2)

ValueError: invalid literal for int() with base 10: 'absent'

<br>Now we can see that we get a second error, a ValueError, because Python can't convert "absent" to an integer. We can add a second try/except to our code:

In [549]:
bonus_scores2 = []
for score in test_scores2:
    try:
        bonus_scores2.append(score + 2)
    except TypeError:
        try:
            bonus_scores2.append(int(score) + 2)
        except ValueError:
            bonus_scores2.append(score)
print(bonus_scores2)

[92, 97, 'absent', 102, 78]


<br>This was a complicated example, but it shows how it can take some logic to get the results you want.

### <br><span style="color:red">Loops Exercise 4

Store this list of students:

In [552]:
students = ["Dustin", "Mike", "Will", "Lucas", 11, "Max"]

<br>Run the code below to identify the error:

In [555]:
for student in students:
    print(len(student))

6
4
4
5


TypeError: object of type 'int' has no len()

<br>Write a new loop to print the length of each student's name. Include a try/except block. You can decide what you want to do with 11 - pass? print something? change to a string?

In [558]:
for student in students:
    try:
        print(len(student))
    except TypeError:
        print(len(str(student)))

6
4
4
5
2
3


Or other options in the except statement

## <br><br><span style="color:teal">2.17 Tuples

You are already familiar with tuples! You just didn't know they were called that. A tuple follows every function call:

In [562]:
round(489289, -3)

489000

<br>(489289, -3) is a tuple.

<br>A tuple is another Python **container**, like a list. An iterator is a collection of items kept in order that can be looped through.
<br><br>A tuple is the same as a list, but it is **immutable**.

<br><br>Tuples are usually designated by parentheses:

In [567]:
my_tuple = (4, 8, 10)
type(my_tuple)

tuple

but, they are technically designated just by having objects separated by commas:

In [570]:
dog_tuple = "beagle", "boxer", "border collie"
type(dog_tuple)

tuple

<br>To make an empty tuple, you have to use parentheses:

In [573]:
empty_t = ()
type(empty_t)

tuple

<br>You can loop through a tuple:

In [576]:
for dog in dog_tuple:
    print(dog)

beagle
boxer
border collie


<br><br>Regular functions will work on tuples:

In [579]:
sum((6, 7, 8))

21

In [581]:
len(dog_tuple)

3

<br>Most list methods do not work with tuples:

In [584]:
for dog in dog_tuple:
    if "collie" in dog:
        empty_t.append()

AttributeError: 'tuple' object has no attribute 'append'

<br>In fact, tuples have very limited methods. Tuples are mostly used behind the scenes in Python, but occasionally you will see one returned to you from a function, so it's good to recognize them.

If you remember from our first lesson in Part 1, there are special arithmetic operators for division.
<br><br>Regular division:

In [588]:
21 / 5

4.2

To return only the whole integer:

In [591]:
21 // 5

4

To return only the remainder:

In [594]:
21 % 5

1

There is also a built-in function to return both the integer and the remainder:

In [597]:
divmod(21,5)

(4, 1)

The result of the function `divmod()` is a tuple!

<br>The function for calculating a T-test for the means of two independent samples returns a fancy tuple with two items. It is found in the **scipy** module.

In [601]:
from scipy.stats import ttest_ind

In [602]:
list_a = [7, 4, 3, 9, 8, 29, 8, 17]
list_b = [8, 10, 14, 28, 26, 18, 17, 17]
answer = ttest_ind(list_a, list_b)

In [603]:
print(answer)

TtestResult(statistic=-1.6991019936393787, pvalue=0.11140195153800304, df=14.0)


<br>This is a very fancy tuple, with keywords, but it is still a tuple.
<br>You can index a tuple to get the info you need:

In [605]:
stat = answer[0]
p_value = answer[1]
print(stat)
print(p_value)         

-1.6991019936393787
0.11140195153800304


<br>**You can always change a tuple to a list to make it easier to work with**:

In [607]:
answer_list = list(answer)
print(answer_list)

[-1.6991019936393787, 0.11140195153800304]


## <br><br><span style="color:teal">2.18 Sets

**Sets are like unordered lists with no duplicate values.**

They are great for removing duplicates from lists:

In [616]:
num_list = [4, 4, 4, 4, 4, 4, 10]
num_set = set(num_list)
print(num_set)

{10, 4}


You can change them back to lists in the same line of code to quickly remove duplicates from a list:

In [619]:
num_list = [4, 4, 4, 4, 4, 4, 10]
num_list = list(set(num_list))
print(num_list)

[10, 4]


<br>Sets are designated by curly brackets (Python knows it's not a dictionary if you don't include any colons inside), or just by using the set`()` function on a list:

In [622]:
s = {8, 9, 10, 9, 8}
print(s)

{8, 9, 10}


In [624]:
s2 = set([8, 9, 10, 9, 8])
print(s2)

{8, 9, 10}


## <br><span style="color:red">Exercise: sets

Store the following list. This is a list of birds you observed from your window.

In [628]:
bird_obs = ["Mourning Dove", 
            "Red-winged Blackbird", 
            "Red-winged Blackbird", 
            "Red-winged Blackbird", 
            "Mourning Dove", 
            "Mourning Dove", 
            "Northern Cardinal", 
            "Mourning Dove", 
            "Song Sparrow", 
            "Song Sparrow", 
            "Song Sparrow", 
            "American Crow",
            "American Robin", 
            "American Crow",
            "Song Sparrow", 
            "Red-winged Blackbird", 
            "Mourning Dove", 
            "Red-winged Blackbird", 
            "Blue Jay", 
            "Red-winged Blackbird", 
            "Song Sparrow", 
            "Red-winged Blackbird", 
            "Mourning Dove", 
            "Red-winged Blackbird", 
            "European Starling"]

How many total birds did you observe? Write code to find the length of the `bird_obs` list:

In [631]:
len(bird_obs)

25

How many different types of birds did you observe? Write code to turn the `bird_obs` list into a set, and then find the length of that set. Bonus if you can do it in one line of code:

In [634]:
len(set(bird_obs))

8

Which different bird types did you see? Turn the `bird_obs` list into a set, and loop through it, printing out each bird type:

In [637]:
for bird in set(bird_obs):
    print(bird)

Blue Jay
Song Sparrow
Mourning Dove
American Crow
Northern Cardinal
Red-winged Blackbird
American Robin
European Starling


## <br><br><span style="color:teal">2.19 Ranges

Another Python collection! A range is created by the `range()` function. It is used to create a collection of integers in order so that you can loop through it.

The `range()` function takes 2 or 3 arguments: 
- which integer to start with
- which integer to end with
- how many integers to iterate by between numbers. *If you do not provide the third argument, the range will iterate by 1.*

<br>If you run the function alone, it will return the range object:

In [644]:
range(1, 10, 2)

range(1, 10, 2)

<br>This isn't very useful. But, if you loop through the range object, you can do something with each number:

In [647]:
for i in range(1, 10, 2):
    print(i)

1
3
5
7
9


<br>**You can also change the range object into a list to work with it**:

In [650]:
r = list(range(1, 10, 2))
print(r)

[1, 3, 5, 7, 9]


*Note: `range()` does not work with floats, only integers. To create a range of floats, you will need to import the module `numpy`.*

<br>Like Python indexing, the stop position is **exclusive** of the number passed as the stop argument:

In [654]:
for i in range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


## <br><span style="color:red">Exercise: range

Write a for loop to print every 20th number from 0 to 400.

In [658]:
for i in range(0, 401, 20):
    print(i)

0
20
40
60
80
100
120
140
160
180
200
220
240
260
280
300
320
340
360
380
400


## <br><br><span style="color:teal">2.20 Dictionaries

Earlier, we practiced looping through fictional characters from Avengers, Star Wars, and Moana. But how did the computer know who was in which movie, or who could fly?

A dictionary is **a collection of *key: value* pairs.** 
- Dictionaries are surrounded by curly brackets {}
- **key: value pairs** inside the dictionary are separated by commas
- In each **key: value pair**, the key and value are separated by a colon :
- The key must always be a string
- The value can be any object

<br>Here's a dictionary of heights, in inches. The keys are people's names and the values are integers:

In [661]:
inches_dict = {"Jo": 60, "Rae": 68, "Tom": 65}

`dict` *is a common abbreviation for a dictionary in Python.*

<br>This dictionary has info about someone's mom. The keys are trait categories and the values are a mix of integers and strings:

In [665]:
mom_dict = {"height": 65, "eyes": "hazel", 
            "hair": "gray", "age": 70}

<br>This dictionary contains results of an experiment. The keys are the names of the test runs and the values are lists of floats:

In [668]:
results = {"test1": [3.4, 0.2, 1.4, 2.2, 8.0], 
           "test2": [0.9, 3.4, 2.5, 4.7, 2.6], 
           "test3": [4.9, 2.4, 0.4, 8.4, 2.5]}

*If a dictionary is long, you can write it on multiple lines, just like a list.*

## <br><br><span style="color:teal">2.21 Indexing a dictionary

In [672]:
grade_dict = {"Charlie": [90, 96, 89, 79], 
              "Tony": [99, 98, 96, 93], 
              "Suman": [85, 88, 83, 87],
              "Yuvie": [66, 76, 80, 62],
              "May": [97, 94, 89, 91]}

print(grade_dict)

{'Charlie': [90, 96, 89, 79], 'Tony': [99, 98, 96, 93], 'Suman': [85, 88, 83, 87], 'Yuvie': [66, 76, 80, 62], 'May': [97, 94, 89, 91]}


<br>**Unlike lists, dictionaries are indexed by the name of the key. They cannot be indexed by position in the dictionary.** In the latest versions of Python, dictionaries are saved in order, but the purpose of a dictionary isn't to keep entries in numerical order - would you ever need to know what the 110th word in the Oxford English Dictionary is?

In [675]:
grade_dict["Tony"]

[99, 98, 96, 93]

In [677]:
grade_dict[3]

KeyError: 3

<br>To index something inside a value, first index the key, then the position in the value. In our `grade_dict` example, the values are lists, so to index Tony's last grade:

In [680]:
grade_dict["Tony"][-1]

93

### <br><span style="color:red">Exercise: Creating and indexing dictionaries

Create a dictionary called `favorites`. The keys should be "color", "food", and "song". The values should be your favorite color, food, and song. 

In [687]:
favorites = {"color": "blue", "food": "strawberries", "song": "The Wheels on the Bus"}

Write code to index your favorite song:

In [692]:
favorites["song"]

'The Wheels on the Bus'

Write code to index the third letter of your favorite food.

In [696]:
favorites["food"][2]

'r'

## <br><br><span style="color:teal">2.22 Adding an entry to a dictionary

You don't have to use a function to add to a dictionary. Just **index** a new key and **assign** it a value:

Let's look at the `grade_dict` as it is now, and then add a new student.

In [701]:
print(grade_dict)

{'Charlie': [90, 96, 89, 79], 'Tony': [99, 98, 96, 93], 'Suman': [85, 88, 83, 87], 'Yuvie': [66, 76, 80, 62], 'May': [97, 94, 89, 91]}


In [703]:
grade_dict["Ben"] = [60, 57, 63]

In [705]:
print(grade_dict)

{'Charlie': [90, 96, 89, 79], 'Tony': [99, 98, 96, 93], 'Suman': [85, 88, 83, 87], 'Yuvie': [66, 76, 80, 62], 'May': [97, 94, 89, 91], 'Ben': [60, 57, 63]}


<br>If the item already exists in the dictionary, you will overwrite it:

In [708]:
grade_dict["Ben"] = [82, 88, 90]
print(grade_dict)

{'Charlie': [90, 96, 89, 79], 'Tony': [99, 98, 96, 93], 'Suman': [85, 88, 83, 87], 'Yuvie': [66, 76, 80, 62], 'May': [97, 94, 89, 91], 'Ben': [82, 88, 90]}


You should always be careful that your dictionary keys are unique.

### <br><span style="color:red">Exercise: Adding to dictionaries

Add a new key:value pair to your `favorites` dictionary. The key could be tv_show, movie, book, or anything else you'd like to add. 

In [714]:
favorites["book"] = "Jamberry"

In [717]:
print(favorites)

{'color': 'blue', 'food': 'strawberries', 'song': 'The Wheels on the Bus', 'book': 'Jamberry'}


## <br><br><span style="color:teal">2.23 Looping through a dictionary

Let's try to loop through a dictionary the same way we would loop through a list:

In [722]:
for entry in grade_dict:
    print(entry)

Charlie
Tony
Suman
Yuvie
May
Ben


If you have an updated version of Python 3, it will print out the keys in the order you gave them when you first created the dictionary. If you have a slightly older version of Python 3, it might give you an error.
<br><br>In either case, this way doesn't allow us to loop through the values or loop through both the keys and the values. 
<br><br>Python has dictionary methods that will give you a list of just the `keys` or just the `values`:

In [725]:
grade_dict.keys()

dict_keys(['Charlie', 'Tony', 'Suman', 'Yuvie', 'May', 'Ben'])

In [727]:
grade_dict.values()

dict_values([[90, 96, 89, 79], [99, 98, 96, 93], [85, 88, 83, 87], [66, 76, 80, 62], [97, 94, 89, 91], [82, 88, 90]])

Our grade_dict dictionary values are lists, so we get a list of lists.

<br>We can loop through a list! Let's be more explicit and tell the computer that we want to loop through only the keys by adding the `keys()` method to the end of our dictionary:

In [731]:
for key in grade_dict.keys():
    print(key)

Charlie
Tony
Suman
Yuvie
May
Ben


Or we can loop through the values:

In [734]:
for value in grade_dict.values():
    print(value)

[90, 96, 89, 79]
[99, 98, 96, 93]
[85, 88, 83, 87]
[66, 76, 80, 62]
[97, 94, 89, 91]
[82, 88, 90]


<br>Remember that we can give our temporary variable any name we want in our for loop. This is commonly used:

In [737]:
for k in grade_dict.keys():
    print(k)

Charlie
Tony
Suman
Yuvie
May
Ben


In [739]:
for v in grade_dict.values():
    print(v)

[90, 96, 89, 79]
[99, 98, 96, 93]
[85, 88, 83, 87]
[66, 76, 80, 62]
[97, 94, 89, 91]
[82, 88, 90]


<br>But it's also good to use more appropriate variable names:

In [742]:
for student in grade_dict.keys():
    print(student)

Charlie
Tony
Suman
Yuvie
May
Ben


In [744]:
for grade_list in grade_dict.values():
    print(grade_list)

[90, 96, 89, 79]
[99, 98, 96, 93]
[85, 88, 83, 87]
[66, 76, 80, 62]
[97, 94, 89, 91]
[82, 88, 90]


### <br><span style="color:red">Exercise: Looping through a dictionary

Write code to loop through the keys of your favorites dictionary. Print out each key.

In [747]:
for i in favorites.keys():
    print(i)

color
food
song
book


<br>Write code to loop through the values of your favorites dictionary. Print out each value.

In [750]:
for v in favorites.values():
    print(v)

blue
strawberries
The Wheels on the Bus
Jamberry


<br><br><br>There is also a method function for returning both the keys and the values:

In [753]:
grade_dict.items()

dict_items([('Charlie', [90, 96, 89, 79]), ('Tony', [99, 98, 96, 93]), ('Suman', [85, 88, 83, 87]), ('Yuvie', [66, 76, 80, 62]), ('May', [97, 94, 89, 91]), ('Ben', [82, 88, 90])])

The `.items()` method returns a list with one tuple for each key-value pair. 
<br><br><br>We can also loop through the `.items()` tuples, but we have to include **two temporary variables** in our `for` loop statement instead of one:

In [756]:
for k, v in grade_dict.items():
    print(k)
    print(v)

Charlie
[90, 96, 89, 79]
Tony
[99, 98, 96, 93]
Suman
[85, 88, 83, 87]
Yuvie
[66, 76, 80, 62]
May
[97, 94, 89, 91]
Ben
[82, 88, 90]


In [758]:
for student, grade_list in grade_dict.items():
    print(student)
    print(grade_list)

Charlie
[90, 96, 89, 79]
Tony
[99, 98, 96, 93]
Suman
[85, 88, 83, 87]
Yuvie
[66, 76, 80, 62]
May
[97, 94, 89, 91]
Ben
[82, 88, 90]


### <br><span style="color:red">Exercise: Looping through a dictionary, both keys and values

Write code to loop through both the keys and values of your favorites dictionary. Print out a full sentence for each favorite, like "My favorite color is purple."

In [762]:
for k, v  in favorites.items():
    print("My favorite " + k + " is " + v + ".")

My favorite color is blue.
My favorite food is strawberries.
My favorite song is The Wheels on the Bus.
My favorite book is Jamberry.


## <br><br><span style="color:teal">2.24 Nested loops
Since our values are list objects, we can also use a **nested loop** to loop through both the dictionary and the grade lists:

In [766]:
for student, grade_list in grade_dict.items():
    print(student)
    for grade in grade_list:
        print(grade)

Charlie
90
96
89
79
Tony
99
98
96
93
Suman
85
88
83
87
Yuvie
66
76
80
62
May
97
94
89
91
Ben
82
88
90


That code is called a **nested loop** - a loop inside a loop!

### <br><span style="color:red">Exercise: Nested loop

Run the cell below to store the `nicknames` dictionary. The keys are the full names, and the values are the nicknames.

In [771]:
nicknames = {"Charles": "Charlie", 
             "Anthony": "Tony", 
             "Suman": "Suman", 
             "Yuval": "Yuvie", 
             "May-Lin": "May", 
             "Benjamin": "Ben",
             "Keesha": "Keesha"}

Write a nested loop to print out each letter in each person's nickname:

In [774]:
for v in nicknames.values():
    for i in v:
        print(i)

C
h
a
r
l
i
e
T
o
n
y
S
u
m
a
n
Y
u
v
i
e
M
a
y
B
e
n
K
e
e
s
h
a


*Hint: You must first decide if you need to loop through the keys, the values, or both. Do you need both to get the job done, or only the keys, or only the values?*

## <br><br><span style="color:teal">2.25 Filtering a dictionary
Just like we did with a list, we will want to combine a for loop with an if statement (or an if/elif/else statement).
<br><br>Let's practice with our `nicknames` dictionary. Some of the nicknames are the same as the full names. Let's loop through the dictionary. If the person has a nickname that is different than their full name, we will print out a sentence like "Catherine goes by Katie." If they do not have a different nickname, we will print out a sentence like "Kelsey does not have a nickname."

In [780]:
nicknames = {"Charles": "Charlie", 
             "Anthony": "Tony", 
             "Suman": "Suman", 
             "Yuval": "Yuvie", 
             "May-Lin": "May", 
             "Benjamin": "Ben",
             "Keesha": "Keesha"}

In [782]:
for k, v in nicknames.items():
    if k != v:
        print(k + " goes by " + v + ".")
    else:
        print(k + " does not have a nickname.")

Charles goes by Charlie.
Anthony goes by Tony.
Suman does not have a nickname.
Yuval goes by Yuvie.
May-Lin goes by May.
Benjamin goes by Ben.
Keesha does not have a nickname.


### <br><br><span style="color:red">Exercise: Filtering dictionaries

Here's an example dictionary. Run the cell below:

In [786]:
hero_dict = {"Captain Marvel": "Avengers", "Finn": "Star Wars", 
             "Maui": "Moana", "Captain America": "Avengers", 
             "Princess Leia": "Star Wars"}

<br>Using `hero_dict`, write a for loop/if statement to print the name of all of the characters in Star Wars: 

In [789]:
for k,v in hero_dict.items():
    if v == "Star Wars":
        print(k)

Finn
Princess Leia


OR

In [792]:
for k,v in hero_dict.items():
    if "Star Wars" in v:
        print(k)

Finn
Princess Leia


## <br><br><span style="color:teal">2.26 Adding key:value pairs to an empty dictionary

Yesterday we learned how to loop through a list and add items to a new empty list. We can also do that with dictionaries.
<br><br>We will work with the grade_dict we used earlier:

In [797]:
grade_dict = {"Charlie": [90, 96, 89, 79], 
              "Tony": [99, 98, 96, 93], 
              "Suman": [85, 88, 83, 87],
              "Yuvie": [66, 76, 80, 62],
              "May": [97, 94, 89, 91],
              "Ben": [82, 88, 90]}

print(grade_dict)

{'Charlie': [90, 96, 89, 79], 'Tony': [99, 98, 96, 93], 'Suman': [85, 88, 83, 87], 'Yuvie': [66, 76, 80, 62], 'May': [97, 94, 89, 91], 'Ben': [82, 88, 90]}


<br><br>**<span style="color:crimson">LOGIC** Here we will create a new dictionary from the data in the grades_dict. For the new dictionary, the keys will be the students' names and the values will be their final score for the class. The final score will be calculated as the mean of all the scores in their grade list. We can use the mean function from the statistics package.

In [800]:
import statistics

<br>First, we create an empty dictionary. Next, we loop through the old dictionary, calculate each person's final grade, and add them to the new dictionary:

In [803]:
final_dict = {}
for student, grade_list in grade_dict.items():
    final_score = statistics.mean(grade_list)
    final_dict[student] = final_score
print(final_dict)

{'Charlie': 88.5, 'Tony': 96.5, 'Suman': 85.75, 'Yuvie': 71, 'May': 92.75, 'Ben': 86.66666666666667}


<br>We can make the dictionary look more consistent by rounding all the numbers to two places after the decimal point:

In [806]:
final_dict = {}
for student, grade_list in grade_dict.items():
    final_score = round(statistics.mean(grade_list), 2)
    final_dict[student] = final_score
print(final_dict)

{'Charlie': 88.5, 'Tony': 96.5, 'Suman': 85.75, 'Yuvie': 71, 'May': 92.75, 'Ben': 86.67}


<br>The round function won't include trailing zeros, like on 96.5. We can make the dictionary look a little more consistent by changing the whole numbers, like 71, to floats:

In [809]:
final_dict = {}
for student, grade_list in grade_dict.items():
    final_score = round(float(statistics.mean(grade_list)), 2)
    final_dict[student] = final_score
print(final_dict)

{'Charlie': 88.5, 'Tony': 96.5, 'Suman': 85.75, 'Yuvie': 71.0, 'May': 92.75, 'Ben': 86.67}


## <br><br><span style="color:teal">2.27 List of dictionaries

Sometimes it is useful to have a list of dictionaries because that is how your data is best represented. You can index individual data points in the list or dictionaries, and you can loop through both levels.

In [813]:
gradebook = [{"name": "Zygon", "HW1": 10, "HW2": 10, "HW3": 10}, 
             {"name": "Vogon", "HW1": 10, "HW2": 10, "HW3": 10}, 
             {"name": "Cylon", "HW1": 10, "HW2": 10, "HW3": 10}, 
             {"name": "Mudokon", "HW1": 7, "HW2": 8, "HW3": 6}]

<br>**Indexing a list of dictionaries**
<br>To return an individual dictionary, you use list indexing because each dictionary is an item in the list. What do you think this will return?

In [816]:
gradebook[2]

{'name': 'Cylon', 'HW1': 10, 'HW2': 10, 'HW3': 10}

<br>To return a value in one of the dictionaries, you first index the dictionary's place in the list, and then index the key in your key:value pair of interest. What do you think this will return?

In [819]:
gradebook[3]["HW1"]

7

<br>**Looping through a list of dictionaries**
<br>**<span style="color:crimson">LOGIC** Let's print out a full sentence for each student. The sentence should report the sum of all the student's homework assignment grades.

It's often useful to first just print each item in a loop, to confirm that you know what you're looking at. I do this all the time when I code:

In [823]:
for dictionary in gradebook:
    print(dictionary)

{'name': 'Zygon', 'HW1': 10, 'HW2': 10, 'HW3': 10}
{'name': 'Vogon', 'HW1': 10, 'HW2': 10, 'HW3': 10}
{'name': 'Cylon', 'HW1': 10, 'HW2': 10, 'HW3': 10}
{'name': 'Mudokon', 'HW1': 7, 'HW2': 8, 'HW3': 6}


Then you can do more and slowly build up your loop:

In [826]:
for dictionary in gradebook:
    name = dictionary["name"]
    print(name)

Zygon
Vogon
Cylon
Mudokon


In [828]:
for dictionary in gradebook:
    name = dictionary["name"]
    HW_total = dictionary["HW1"] + dictionary["HW2"] + dictionary["HW3"]
    print(HW_total)

30
30
30
21


In [830]:
for dictionary in gradebook:
    name = dictionary["name"]
    HW_total = dictionary["HW1"] + dictionary["HW2"] + dictionary["HW3"]
    print(name + " scored " + str(HW_total) + " points on Homework")

Zygon scored 30 points on Homework
Vogon scored 30 points on Homework
Cylon scored 30 points on Homework
Mudokon scored 21 points on Homework


## <br><br><span style="color:teal">2.28 Hard coding
The code we just wrote is ok, but it wouldn't work if more than 3 homework assignments were added. Let's say you teach the same class next year and you want to reuse the code, only next year you give 4 homework assignments instead of 3. 

<br>When there are details in the code specific to your data, we say they are **hard coded**.
<br><br>As a beginner, you will do a lot of hard coding to solve your problems, but if you ever want to reuse your scripts or share them with someone else, you will need to try to not hard code.

<br>Instead, you can loop through the list and then loop through the dictionary (a **nested loop**). 
<br><br>I've included **comments** in the code to explain what I'm doing. Comments start with a `#` and are ignored by the computer.

In [835]:
for dictionary in gradebook: #loop through the list
    name = dictionary["name"] #get student's name
    HW_total = 0 
    for k, v in dictionary.items(): #loop through the key:value pairs
        if k != "name": #get every key:value pair except name
            HW_total = HW_total + v #add the value to our HW total
    print(name + " scored " + str(HW_total) + " points on Homework")

Zygon scored 30 points on Homework
Vogon scored 30 points on Homework
Cylon scored 30 points on Homework
Mudokon scored 21 points on Homework


<br>**OR**<br>

In [838]:
for dictionary in gradebook: #loop through the list
    name = dictionary["name"] #get student's name
    HW_total = 0 
    for v in dictionary.values(): #loop through the values only
        if type(v) != str: #get every value except name, which is a string
            HW_total = HW_total + v #add the value to our HW total
    print(name + " scored " + str(HW_total) + " points on Homework")

Zygon scored 30 points on Homework
Vogon scored 30 points on Homework
Cylon scored 30 points on Homework
Mudokon scored 21 points on Homework


## <br><br><span style="color:teal">2.29 Dictionary of dictionaries

You can also format your data as a dictionary of dictionaries.

### <br><span style="color:red">Exercise: Indexing a nested dictionary

In [843]:
gradebook_dict = {"Zygon": {"HW1": 3, "HW2": 2, "HW3": 4}, 
                  "Vogon": {"HW1": 10, "HW2": 10, "HW3": 10}, 
                  "Cylon": {"HW1": 10, "HW2": 10, "HW3": 10}, 
                  "Mudokon": {"HW1": 7, "HW2": 8, "HW3": 6}}

Use dictionary indexing to write code to return all of Cylon's grades:

In [846]:
gradebook_dict["Cylon"]

{'HW1': 10, 'HW2': 10, 'HW3': 10}

Use dictionary indexing to write code to return Vogon's score on HW2:

In [848]:
gradebook_dict["Vogon"]["HW2"]

10